In [3]:
import os
os.environ["KERAS_BACKEND"] = "torch"
import torch
print("CUDA available:", torch.cuda.is_available())
print("GPU:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else None)

CUDA available: True
GPU: NVIDIA GeForce RTX 4070 Laptop GPU


In [ ]:
import torch, time

device = "cuda" if torch.cuda.is_available() else "cpu"
x = torch.randn(4096, 4096, device=device)
y = torch.randn(4096, 4096, device=device)

t0 = time.time()
z = x @ y
if device == "cuda":
    torch.cuda.synchronize()
print("device:", device, "secs:", time.time() - t0)

# Backend-agnostic test

In [1]:
import os
STEPS = 50

In [2]:
os.environ["KERAS_BACKEND"] = "torch" # 🔁 CHANGE THIS PER ENV: "tensorflow" | "torch" | "jax"
import torch
print("CUDA available:", torch.cuda.is_available())
print("GPU:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else None)

import time
import keras
import numpy as np

print("Backend:", keras.backend.backend())

# -------------------------
# Benchmark configuration
# -------------------------
BATCH_SIZE = 1024
INPUT_DIM = 2048
HIDDEN_DIM = 4096
OUTPUT_DIM = 10
DTYPE = "float32"

# -------------------------
# Generate data
# -------------------------
x = np.random.randn(BATCH_SIZE, INPUT_DIM).astype(DTYPE)
y = np.random.randint(0, OUTPUT_DIM, size=(BATCH_SIZE,))

# -------------------------
# Build model
# -------------------------
model = keras.Sequential(
    [
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(OUTPUT_DIM),
    ]
)

model.compile(
    optimizer="adam",
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
)

# -------------------------
# Warm-up (important!)
# -------------------------
model.train_on_batch(x, y)

# -------------------------
# Timed run
# -------------------------
start = time.perf_counter()

for _ in range(STEPS):
    model.train_on_batch(x, y)

elapsed = time.perf_counter() - start

print(f"Total time ({STEPS} steps): {elapsed:.3f} s")
print(f"Time per step: {elapsed / STEPS:.4f} s")


CUDA available: True
GPU: NVIDIA GeForce RTX 4070 Laptop GPU
Backend: torch
Total time (50 steps): 1.509 s
Time per step: 0.0302 s


In [2]:
os.environ["KERAS_BACKEND"] = "torch" # 🔁 CHANGE THIS PER ENV: "tensorflow" | "torch" | "jax"
import torch
print("CUDA available:", torch.cuda.is_available())
print("GPU:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else None)

import time
import keras
import numpy as np

print("Backend:", keras.backend.backend())

# -------------------------
# Benchmark configuration
# -------------------------
BATCH_SIZE = 1024
INPUT_DIM = 2048
HIDDEN_DIM = 4096
OUTPUT_DIM = 10
DTYPE = "float32"

# -------------------------
# Generate data
# -------------------------
x = np.random.randn(BATCH_SIZE, INPUT_DIM).astype(DTYPE)
y = np.random.randint(0, OUTPUT_DIM, size=(BATCH_SIZE,))

# -------------------------
# Build model
# -------------------------
model = keras.Sequential(
    [
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(OUTPUT_DIM),
    ]
)

model.compile(
    optimizer="adam",
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
)

# -------------------------
# Warm-up (important!)
# -------------------------
model.train_on_batch(x, y)

# -------------------------
# Timed run
# -------------------------
start = time.perf_counter()

for _ in range(STEPS):
    model.train_on_batch(x, y)

elapsed = time.perf_counter() - start

print(f"Total time ({STEPS} steps): {elapsed:.3f} s")
print(f"Time per step: {elapsed / STEPS:.4f} s")


CUDA available: False
GPU: None
Backend: torch
Total time (50 steps): 13.020 s
Time per step: 0.2604 s


In [4]:
os.environ["KERAS_BACKEND"] = "tensorflow" # 🔁 CHANGE THIS PER ENV: "tensorflow" | "torch" | "jax"

import time
import keras
import numpy as np

print("Backend:", keras.backend.backend())

# -------------------------
# Benchmark configuration
# -------------------------
BATCH_SIZE = 1024
INPUT_DIM = 2048
HIDDEN_DIM = 4096
OUTPUT_DIM = 10
DTYPE = "float32"

# -------------------------
# Generate data
# -------------------------
x = np.random.randn(BATCH_SIZE, INPUT_DIM).astype(DTYPE)
y = np.random.randint(0, OUTPUT_DIM, size=(BATCH_SIZE,))

# -------------------------
# Build model
# -------------------------
model = keras.Sequential(
    [
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(OUTPUT_DIM),
    ]
)

model.compile(
    optimizer="adam",
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
)

# -------------------------
# Warm-up (important!)
# -------------------------
model.train_on_batch(x, y)

# -------------------------
# Timed run
# -------------------------
start = time.perf_counter()

for _ in range(STEPS):
    model.train_on_batch(x, y)

elapsed = time.perf_counter() - start

print(f"Total time ({STEPS} steps): {elapsed:.3f} s")
print(f"Time per step: {elapsed / STEPS:.4f} s")


Backend: tensorflow
Total time (50 steps): 8.805 s
Time per step: 0.1761 s


In [2]:
os.environ["KERAS_BACKEND"] = "jax" # 🔁 CHANGE THIS PER ENV: "tensorflow" | "torch" | "jax"

import time
import keras
import numpy as np

print("Backend:", keras.backend.backend())

# -------------------------
# Benchmark configuration
# -------------------------
BATCH_SIZE = 1024
INPUT_DIM = 2048
HIDDEN_DIM = 4096
OUTPUT_DIM = 10
DTYPE = "float32"

# -------------------------
# Generate data
# -------------------------
x = np.random.randn(BATCH_SIZE, INPUT_DIM).astype(DTYPE)
y = np.random.randint(0, OUTPUT_DIM, size=(BATCH_SIZE,))

# -------------------------
# Build model
# -------------------------
model = keras.Sequential(
    [
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(HIDDEN_DIM, activation="relu"),
        keras.layers.Dense(OUTPUT_DIM),
    ]
)

model.compile(
    optimizer="adam",
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
)

# -------------------------
# Warm-up (important!)
# -------------------------
model.train_on_batch(x, y)

# -------------------------
# Timed run
# -------------------------
start = time.perf_counter()

for _ in range(STEPS):
    model.train_on_batch(x, y)

elapsed = time.perf_counter() - start

print(f"Total time ({STEPS} steps): {elapsed:.3f} s")
print(f"Time per step: {elapsed / STEPS:.4f} s")


Backend: jax
Total time (50 steps): 6.972 s
Time per step: 0.1394 s


# Benchmark Summary: Keras Backend Performance (50 steps)

| Backend | Device | GPU Model | Total Time | Time per Step | Slowdown vs Fastest |
|---------|--------|-----------|------------|---------------|---------------------|
| **PyTorch** | GPU | RTX 4070 Laptop | 1.509 s | 0.0302 s | **1.0× (baseline)** |
| **JAX** | CPU* | N/A | 6.972 s | 0.1394 s | 4.6× |
| **TensorFlow** | CPU* | N/A | 8.805 s | 0.1761 s | 5.8× |
| **PyTorch** | CPU | None | 13.020 s | 0.2604 s | 8.6× |

\* *Assuming CPU execution based on the absence of GPU info in the output*

## Key Findings

- **GPU Acceleration**: PyTorch with CUDA on RTX 4070 is **8.6× faster** than PyTorch on CPU
- **Backend Comparison (CPU)**: JAX (6.972s) > TensorFlow (8.805s) > PyTorch (13.020s)
- **Fastest Overall**: PyTorch + GPU (1.509s total, 0.0302s per step)
- **CPU Performance**: JAX is the fastest CPU backend, ~1.9× faster than PyTorch CPU and ~1.3× faster than TensorFlow

The RTX 4070 GPU provides substantial acceleration, being **4.6× faster** than even the most optimized CPU backend (JAX).